### ISSUE TITLE
Console Action Refresh Not Writing Files to Server Using Blitz File Storage

### ISSUE THREAD

**Issue by curtismorte** [putyourlightson/craft-blitz#437]
**The plugin and Craft version numbers.**
* Craft: `4.1.4.1`
* Blitz: `4.2.0`

**Description of the issue:**
When running the console command `blitz/cache/refresh` my expectation is that upon completion, I should be able to find all of my generated html files according to the settings for my cache storage. Currently, I am not seeing any files being written to the server with "Blitz File Storage" as my selected cache storage. I know that the cache storage works because when I visit a page I am able to see the cache.
> I have explicitly made sure that the alias for `@web` is defined and my webroot is specified to be `public` as per the documentation [here](https://putyourlightson.com/plugins/blitz#the-site-is-not-being-cached-when-using-console-commands)

**Here are my project configuration settings:**
```
blitz:
    edition: standard
    enabled: true
    licenseKey: **********************
    schemaVersion: 4.2.0
    settings:
      apiKey: ''
      cacheControlHeader: 'public, s-maxage=31536000, max-age=0'
      cacheDuration: null
      cacheElementQueries: true
      cacheElements: true
      cacheGeneratorSettings:
        __assoc__:
          -
            - concurrency
            - '3'
      cacheGeneratorType: putyourlightson\blitz\drivers\generators\LocalGenerator
      cacheNonHtmlResponses: false
      cachePurgerType: putyourlightson\blitz\drivers\purgers\DummyPurger
      cacheStorageSettings:
        __assoc__:
          -
            - folderPath
            - '@webroot/cache/blitz'
          -
            - createGzipFiles
            - '1'
          -
            - createBrotliFiles
            - ''
      cacheStorageType: putyourlightson\blitz\drivers\storage\FileStorage
      cachingEnabled: true
      debug: false
      deployerType: putyourlightson\blitz\drivers\deployers\DummyDeployer
      driverJobPriority: 100
      excludedQueryStringParams:
        -
          __assoc__:
            -
              - queryStringParam
              - gclid
        -
          __assoc__:
            -
              - queryStringParam
              - fbclid
        -
          __assoc__:
            -
              - queryStringParam
              - msclkid
        -
          __assoc__:
            -
              - queryStringParam
              - utm_source
        -
          __assoc__:
            -
              - queryStringParam
              - utm_medium
        -
          __assoc__:
            -
              - queryStringParam
              - utm_term
        -
          __assoc__:
            -
              - queryStringParam
              - utm_campaign
        -
          __assoc__:
            -
              - queryStringParam
              - utm_content
      excludedUriPatterns:
        -
          __assoc__:
            -
              - uriPattern
              - menu
        -
          __assoc__:
            -
              - uriPattern
              - ccpa
      generatePagesWithQueryStringParams: true
      hintsEnabled: true
      includedQueryStringParams:
        -
          __assoc__:
            -
              - queryStringParam
              - '.*'
      includedUriPatterns:
        -
          __assoc__:
            -
              - uriPattern
              - '.*'
      injectScriptEvent: DOMContentLoaded
      integrations:
        - putyourlightson\blitz\drivers\integrations\CommerceIntegration
        - putyourlightson\blitz\drivers\integrations\FeedMeIntegration
        - putyourlightson\blitz\drivers\integrations\SeomaticIntegration
      maxRetryAttempts: 10
      mutexTimeout: 1
      outputComments: true
      queryStringCaching: 0
      queueJobTtr: 300
      refreshCacheAutomaticallyForGlobals: true
      refreshCacheJobPriority: 10
      refreshCacheWhenElementMovedInStructure: true
      refreshCacheWhenElementSavedNotLive: false
      refreshCacheWhenElementSavedUnchanged: false
      refreshMode: 3
      sendPoweredByHeader: true
```

So I took it upon myself to dive into this issue a little further. What I have found is that the old v3 of the plugin will eventually trace down to [this point](https://github.com/putyourlightson/craft-blitz/blob/fe9da1b97363c1d713da2a347bbe70529feddcad/src/drivers/warmers/LocalWarmer.php#L170) when calling the v3 warm action (which is now generate in v4). Thus, the file is saved according to the cache storage. However, in v4, you've update the action to become generate (from warm), and you have rewritten the implementation using AMPHP. This seems to be a much better implementation, but I am not seeing the response data being saved [here](https://github.com/putyourlightson/craft-blitz/blob/4.2.0/src/drivers/generators/LocalGenerator.php#L108) like it was in the old v3 implementation. It seems that the new PHP Generator defined in [local-generator.php](https://github.com/putyourlightson/craft-blitz/blob/4.2.0/src/drivers/generators/local-generator-script.php) isn't actually saving anything either. 

Did the save method get removed accidentally, or was this intentional?

**Comment by bencroker** [comment]
Thanks for the detailed issue! The Local Generator does work in Blitz 4, so this is likely an environment specific issue. Can you enable the `debug` config setting (https://putyourlightson.com/plugins/blitz#debugging) and run the `blitz/cache/generate` command to see if any log errors appear?

**Comment by curtismorte** [comment]
Thanks for the response @bencroker.

I'm able to verify that Local Generator does work with Blitz 4. However, I am only able to get it to generate locally when I visit a routable entry, and not from a console command.

I've had debug enabled the whole time, here is the log output from `blitz/cache/refresh`:
```
[2022-07-18 16:01:04] Page not cached because it matches an excluded URI pattern. [https://domain.nitro/menu/consulting-services/]
[2022-07-18 16:01:04] Page not cached because it matches an excluded URI pattern. [https://domain.nitro/menu/supported-platforms/]
[2022-07-18 16:01:04] Page not cached because it matches an excluded URI pattern. [https://domain.nitro/menu/company/]
[2022-07-18 16:01:04] Page not cached because it matches an excluded URI pattern. [https://domain.nitro/ccpa/]
[2022-07-18 16:01:04] Blitz cache successfully cleared.
[2022-07-18 16:01:04] Blitz cache successfully flushed.
[2022-07-18 16:01:04] Blitz cache generation complete.
```
> Menu is coming from Verbb's Navigation plugin. 

Here is the terminal output:
```
Blitz cache successfully cleared.
Blitz cache successfully flushed.
Cache purging is disabled.
Generating Blitz cache...
[=======================================================================================================================] 100% (127/127) ETA: 1 sec.
Blitz cache generation complete.
Deploying is disabled.
```
 
**Details on the specific environment that are relevant for this issue are:**

`.env`
```
# Force errors to be thrown when depreciated calls for craft are used.
HARD_MODE=true

# The environment Craft is currently running in (dev, staging, production, etc.)
ENVIRONMENT="dev"
PRIMARY_SITE_URL="https://domain.nitro"
```

`/config/general.php`:
```
return [
    // Global settings
    '*' => [
        // Dev Mode (see https://craftcms.com/guides/what-dev-mode-does)
        'devMode' => false,

        ...

        'aliases' => [
            // The URL to the folder that contains the index.php file that was loaded for the request
            '@web' => getenv('PRIMARY_SITE_URL'),
            '@rootUrl' => getenv('PRIMARY_SITE_URL'),
        ],

        // Don't enable template caching since we use Blitz
        'enableTemplateCaching' => false,

        // Whether asset URLs should be revved so browsers don’t load cached versions when they’re modified.
        'revAssetUrls' => true,
    ],

    // Dev environment settings
    'dev' => [
        // Dev Mode (see https://craftcms.com/guides/what-dev-mode-does)
        'devMode' => true,

        // Prevent crawlers from indexing pages and following links
        'disallowRobots' => true,

        // Set this to `false` to prevent administrative changes from being made on Staging
        'allowAdminChanges' => true,
    ],

    // Staging environment settings
    'staging' => [
        // Set this to `false` to prevent administrative changes from being made on Staging
        'allowAdminChanges' => true,

        // Don’t allow updates on Staging
        'allowUpdates' => false,

        // Prevent crawlers from indexing pages and following links
        'disallowRobots' => true,

        // Rely on our systemd service for queueing
        'runQueueAutomatically' => false,

        // A unique ID representing the current build of the codebase.
        'buildId' => App::env('DEPLOYMENT_ID'),
    ],

    // Production environment settings
    'production' => [
        // Set this to `false` to prevent administrative changes from being made on Production
        'allowAdminChanges' => false,

        // Don't allow updates on Production
        'allowUpdates' => false,

        // Prevent crawlers from indexing pages and following links
        'disallowRobots' => false,

        // Rely on our systemd service for queueing
        'runQueueAutomatically' => false,

        // A unique ID representing the current build of the codebase.
        'buildId' => App::env('DEPLOYMENT_ID'),
    ],
```

This specific issue is happening on 5 different environments of the same Craft install (4 nitro, 1 stage/production in AWS). So either our environments are that consistent that we have the same issue (like you mentioned), or the issue might be code related.

**Per my original comment, I had noticed that the way you're saving the files differs between the v3 and the v4 of the plugin:**
`v4`: https://github.com/putyourlightson/craft-blitz/blob/4.2.0/src/drivers/generators/LocalGenerator.php#L108
`v3`: https://github.com/putyourlightson/craft-blitz/blob/fe9da1b97363c1d713da2a347bbe70529feddcad/src/drivers/warmers/LocalWarmer.php#L170
I think you've just moved this somewhere else, but I didn't track this down. Could you point me to where you're saving the files locally in the codebase?

I've looked into the codebase, and this is what I am seeing:
* Calling the controller action for refresh calls: https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/console/controllers/CacheController.php#L392
* _generateCache calls the local generator: https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/drivers/generators/LocalGenerator.php#L54
* generateUris then generates the URIs: https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/drivers/generators/LocalGenerator.php#L82
* generateUrisWithProgress then calls the concurrent iterator: https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/drivers/generators/local-generator-script.php#L24

However, I am not seeing the local generator or the local generator script actually save the file to the local filesystem.

Am I missing something here?

**Comment by bencroker** [comment]
The local generator doesn't save the cached page itself, instead it replicates a web request to the URL and runs the Craft app, allowing Blitz to be triggered and then cache the page.

https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/drivers/generators/local-generator-script.php#L98

So the saving still happens here:

https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/services/GenerateCacheService.php#L385

My local tests are passing, so perhaps it is a write permission issue on your end.

Here is where the file is written to, it would be interesting to see if this line is even called. If it fails with an exception then an error should be written to the log file.

https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/drivers/storage/FileStorage.php#L104

**Comment by curtismorte** [comment]
@bencroker thanks for the information. I had inspected the v3 path, but not the v4 path beyond the concurrent connections to mock legit site requests.

If you track your way back through the code starting at this point (the same as you linked to above):
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/services/GenerateCacheService.php#L385

**You get the following:**
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/services/GenerateCacheService.php#L356
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/services/CacheRequestService.php#L342
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/Blitz.php#L271

Logging shows that the following event is being run:
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/Blitz.php#L271-L305

However, the following event isn't being run:
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/Blitz.php#L295-L301

**EDIT:**
I was able to verify that the console command is triggering the event `EVENT_AFTER_PREPARE` from Craft's `craft\elements\db\ElementQuery` coming from `afterPrepare()` (line 1875). However, the event that's defined here isn't being run:
https://github.com/putyourlightson/craft-blitz/blob/154d21a5c80c80000dac3d0f73e88dd71bd04187/src/Blitz.php#L295-L301

Can you verify that you're able to have the event for after prepare is running on your side? This would be an indicator of an environment issue. I'm going to look into the events being dispatched from Yii to see more into what's going on here.

**Comment by bencroker** [comment]
Yes, it is.

Can you try disabling all other modules/plugins and retesting? If it works, then that would indicate that some custom behaviour is preventing the `Response::EVENT_AFTER_PREPARE` event from firing.

**Comment by curtismorte** [comment]
@bencroker working on it. Will get down to the source of the issues soon. Thank you.

**Comment by curtismorte** [comment]
I can confirm that this is being caused by a custom module specific to the codebase that is used across all environments. @bencroker thank you for following along, providing responses, and being beyond responsive. 

The issue specifically is coming from registering a twig extension. In our module's `init()` method, we have:
```php
if(Craft::$app->request->getIsSiteRequest()){
    $this->registerTwigExtensions();
}
```
Which is defined here:
```php
    /**
     * Responsible for registering twig extensions
     * @return void
     */
    private function registerTwigExtensions(): void
    {
        $extensions = [
           HelperExtensions::class
        ];

        foreach ($extensions as $extension) {
            Craft::$app->view->registerTwigExtension(new $extension);
        }
    }
```

And the helper extension is defined like so:
```php
namespace modules\domainmodule\twigextensions;
use Twig\Extension\AbstractExtension;
use Twig\TwigFilter;


/**
 * Filter for strings containing HTML tags. Checks if text (without tags) is not too long, if it is, makes it shorter without breaking the HTML
 */
class HelperExtensions extends AbstractExtension {

    /**
     * Responsible for making text shorter (if it is needed)
     * @param $text - string with text (can contain HTML tags
     * @return string - output string
     */
    public function appendPeriod ($text): string
    {
        $outputText = $text;
        if ($outputText[-1] != '.') {
            $outputText .= '.';
        }
        return $outputText;
    }

    public function getFilters(): array
    {
        return [
            new TwigFilter('appendPeriod', [$this, 'appendPeriod'])
        ];
    }
}
```

We're creating and registering our twig extensions per the specification from [Craft](https://craftcms.com/docs/4.x/extend/extending-twig.html#register-a-twig-extension). To be safe, I've also double checked how to properly build Twig Extensions from [Twig](https://twig.symfony.com/doc/3.x/advanced.html#creating-an-extension). I've also looked at multiple plugin vendor's code, and their twig filter implementations seem to be the same.

I've dug into the twig filters to see if something was wrong, but it seems that including the filter is what is causing the problem.

At the moment, I'm not sure why this is being caused. I would think that other people would have the problem, but I guess registering custom twig filters isn't that popular (or somehow we did it wrong).

**Comment by bencroker** [comment]
Thanks for the details. I replicated your setup and I believe I've managed to fix the issue in https://github.com/putyourlightson/craft-blitz/commit/d9b581c9d83d55b9665476cfda590e0cd901a503, can you please test that this resolves caching the pages and that your `appendPeriod` filter is still working in those same pages?

**Comment by curtismorte** [comment]
@bencroker you have resolved the issue.

**Comment by bencroker** [comment]
Thanks for your help in resolving this!! Released in version 4.2.1.
